File: PyDispatch/mobile_app/lib/database/connection.dart
/// PyDispatch Mobile – MySQL-Datenbankverbindung.
/// Singleton-Pattern für die Verbindung zur gemeinsamen Datenbank.
import 'package:mysql_client/mysql_client.dart';

class DatabaseConnection {
  static DatabaseConnection? _instance;
  MySQLConnection? _connection;

  DatabaseConnection._();

  static DatabaseConnection get instance {
    _instance ??= DatabaseConnection._();
    return _instance!;
  }

  bool get isConnected => _connection != null;

  /// Verbindung herstellen.
  Future<bool> connect({
    required String host,
    required int port,
    required String user,
    required String password,
    required String database,
  }) async {
    try {
      await disconnect();
      _connection = await MySQLConnection.createConnection(
        host: host,
        port: port,
        userName: user,
        password: password,
        databaseName: database,
        secure: false,
      );
      await _connection!.connect();
      return true;
    } catch (e) {
      _connection = null;
      return false;
    }
  }

  /// Verbindung testen (kurzfristig verbinden und trennen).
  Future<(bool, String)> testConnection({
    required String host,
    required int port,
    required String user,
    required String password,
    required String database,
  }) async {
    try {
      final conn = await MySQLConnection.createConnection(
        host: host,
        port: port,
        userName: user,
        password: password,
        databaseName: database,
        secure: false,
      );
      await conn.connect();
      await conn.close();
      return (true, 'Verbindung erfolgreich.');
    } catch (e) {
      return (false, 'Verbindungsfehler: $e');
    }
  }

  /// SELECT-Abfrage ausführen.
  Future<List<Map<String, dynamic>>> execute(
    String query, [
    Map<String, dynamic>? params,
  ]) async {
    try {
      if (_connection == null) return [];
      final result = await _connection!.execute(query, params ?? {});
      final rows = <Map<String, dynamic>>[];
      for (final row in result.rows) {
        rows.add(row.typedAssoc());
      }
      return rows;
    } catch (e) {
      return [];
    }
  }

  /// INSERT/UPDATE/DELETE ausführen.
  Future<bool> executeModify(
    String query, [
    Map<String, dynamic>? params,
  ]) async {
    try {
      if (_connection == null) return false;
      await _connection!.execute(query, params ?? {});
      return true;
    } catch (e) {
      return false;
    }
  }

  /// INSERT ausführen und die neue ID zurückgeben.
  Future<int?> executeInsert(
    String query, [
    Map<String, dynamic>? params,
  ]) async {
    try {
      if (_connection == null) return null;
      final result = await _connection!.execute(query, params ?? {});
      return result.lastInsertID.toInt();
    } catch (e) {
      return null;
    }
  }

  /// Verbindung trennen.
  Future<void> disconnect() async {
    try {
      await _connection?.close();
    } catch (_) {}
    _connection = null;
  }
}

/// Globaler Zugriff.
final db = DatabaseConnection.instance;

File: PyDispatch/leitstelle/database/connection.py
"""
PyDispatch Leitstelle – Datenbank-Verbindungsverwaltung.
Verwaltet die MySQL-Verbindung als Singleton.
"""

import mysql.connector
from mysql.connector import Error


class DatabaseConnection:
    """Singleton-Klasse für die MySQL-Datenbankverbindung."""

    _instance = None
    _connection = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    def connect(self, host: str, port: int, user: str, password: str, database: str) -> bool:
        """Stellt eine Verbindung zur MySQL-Datenbank her."""
        try:
            self._connection = mysql.connector.connect(
                host=host,
                port=port,
                user=user,
                password=password,
                database=database,
                autocommit=False,
                charset="utf8mb4",
                collation="utf8mb4_unicode_ci",
            )
            return True
        except Error as e:
            print(f"[DB] Verbindungsfehler: {e}")
            return False

    def test_connection(self, host: str, port: int, user: str, password: str,
                        database: str = None) -> tuple[bool, str]:
        """Testet die Verbindung ohne sie dauerhaft zu speichern."""
        try:
            kwargs = dict(host=host, port=port, user=user, password=password)
            if database:
                kwargs["database"] = database
            conn = mysql.connector.connect(**kwargs)
            conn.close()
            return True, "Verbindung erfolgreich"
        except Error as e:
            return False, str(e)

    @property
    def connection(self):
        """Gibt die aktive Verbindung zurück."""
        if self._connection and self._connection.is_connected():
            return self._connection
        return None

    def get_cursor(self, dictionary=True):
        """Gibt einen neuen Cursor zurück."""
        if self.connection:
            return self.connection.cursor(dictionary=dictionary)
        raise ConnectionError("Keine aktive Datenbankverbindung.")

    def execute(self, query: str, params: tuple = None) -> list:
        """Führt eine SELECT-Query aus und gibt die Ergebnisse zurück."""
        cursor = self.get_cursor()
        try:
            cursor.execute(query, params)
            results = cursor.fetchall()
            return results
        finally:
            cursor.close()

    def execute_modify(self, query: str, params: tuple = None) -> int:
        """Führt eine INSERT/UPDATE/DELETE-Query aus."""
        cursor = self.get_cursor()
        try:
            cursor.execute(query, params)
            self._connection.commit()
            return cursor.rowcount
        except Error:
            self._connection.rollback()
            raise
        finally:
            cursor.close()

    def execute_insert(self, query: str, params: tuple = None) -> int:
        """Führt ein INSERT aus und gibt die lastrowid zurück."""
        cursor = self.get_cursor()
        try:
            cursor.execute(query, params)
            self._connection.commit()
            return cursor.lastrowid
        except Error:
            self._connection.rollback()
            raise
        finally:
            cursor.close()

    def disconnect(self):
        """Trennt die Datenbankverbindung."""
        if self._connection and self._connection.is_connected():
            self._connection.close()
            self._connection = None

    def is_connected(self) -> bool:
        """Prüft, ob eine aktive Verbindung besteht."""
        return self._connection is not None and self._connection.is_connected()


# Globale Instanz
db = DatabaseConnection()
